keel-core/src/main/kotlin/com/netflix/spinnaker/keel/core/api/ClusterDeployStrategy.kt

```diff
@@ -23,42 +23,90 @@ import java.time.Duration.ZERO
   Type(Highlander::class)
 )
 sealed class ClusterDeployStrategy {
-  abstract fun toOrcaJobProperties(): Map<String, Any?>
   @get:JsonIgnore
   open val isStaggered: Boolean = false
   @get:JsonInclude(NON_EMPTY)
   open val stagger: List<StaggeredRegion> = emptyList()
+  abstract fun toOrcaJobProperties(): Map<String, Any?>
+  abstract fun withDefaultsOmitted(): ClusterDeployStrategy
 }
 
 @JsonTypeName("red-black")
 data class RedBlack(
-  val rollbackOnFailure: Boolean = true,
-  val resizePreviousToZero: Boolean = false,
-  val maxServerGroups: Int = 2,
-  val delayBeforeDisable: Duration = ZERO,
-  val delayBeforeScaleDown: Duration = ZERO,
+  val rollbackOnFailure: Boolean? = true,
+  val resizePreviousToZero: Boolean? = false,
+  val maxServerGroups: Int? = 2,
+  val delayBeforeDisable: Duration? = ZERO,
+  val delayBeforeScaleDown: Duration? = ZERO,
   // The order of this list is important for pauseTime based staggers
   @JsonInclude(NON_EMPTY)
   override val stagger: List<StaggeredRegion> = emptyList()
 ) : ClusterDeployStrategy() {
+
+  companion object {
+    fun fromOrcaStageContext(context: Map<String, Any?>) =
+      RedBlack(
+        rollbackOnFailure = context["rollback"]
+          ?.let { it as Map<String, Any> }
+          ?.get("onFailure") as Boolean,
+        resizePreviousToZero = context["scaleDown"] as Boolean,
+        maxServerGroups = context["maxRemainingAsgs"] as Int,
+        delayBeforeDisable = Duration.ofSeconds((context["delayBeforeDisableSec"] as Int).toLong()),
+        delayBeforeScaleDown = Duration.ofSeconds((context["delayBeforeScaleDownSec"] as Int).toLong())
+      )
+
+    val DEFAULTS = RedBlack()
+  }
+
   override fun toOrcaJobProperties() = mapOf(
     "strategy" to "redblack",
     "maxRemainingAsgs" to maxServerGroups,
-    "delayBeforeDisableSec" to delayBeforeDisable.seconds,
-    "delayBeforeScaleDownSec" to delayBeforeScaleDown.seconds,
+    "delayBeforeDisableSec" to delayBeforeDisable?.seconds,
+    "delayBeforeScaleDownSec" to delayBeforeScaleDown?.seconds,
     "scaleDown" to resizePreviousToZero,
     "rollback" to mapOf("onFailure" to rollbackOnFailure)
   )
 
   override val isStaggered: Boolean
-    get() = stagger.isNotEmpty()
+    get() = stagger?.isNotEmpty() ?: false
+
+  override fun withDefaultsOmitted() =
+    RedBlack(
```

Comment:
I tried to use my `ResourceSpecExportHelper` to nullify defaults on these objects, like we do for other types, but the challenge is that, in this case, we want the type-identifying field `strategy` to always be present, even if the other fields are all `null`. After spending an hour trying to fiddle with that code, I gave up and went for the simple thing that works, even though it hurts my 👀 to look at it. Oh well. 🤷‍♂️